### ISSUE TITLE
[POEM] Action Recognition

### ISSUE THREAD

**Issue by yuan0101213** [google-research/google-research#611]
Thanks for sharing of your work .
In the paper(View-Invariant Probabilistic Embedding forHuman Pose),the embeddings for action recogni-
tion is evaluated by using nearest neighbor search with the sequence distance . 
Can I understand in this way that some pose have already been recognized, and other pose use nearest neighbor search with the sequence distance to find the nearest already recognized pose?Does the model need to additional training?

**Comment by jenjsun** [comment]
Hello, and thanks for your interest! This is described in Section 4.3 of our paper and here are the steps you can take to do action recognition:
1. There is no additional training of the embedding model (Pr-VIPE).
2. We compute Pr-VIPE embeddings on the detected pose from each video frame in Penn Action. To be more precise, we used detected poses given the person boxes provided by the Penn Action and for frames without any given box (box is missing), we smear a box from a closest frame.
3. The frame matching distance for a pair of frames i, j is the average of the the neg log of the matching probability of frames (i-6, i-3, i, i+3, i+6) and (j-6, j-3, j, j+3, j+6), of each corresponding frame pair
4. Use DTW to align two videos using frame matching distance, and the nearest neighbor is the video with the smallest averaged distance  after alignment
5. Do action classification of an unlabelled video by finding the nearest neighbor of a labelled video using the procedure above
There are additional implementation details in the Supplementary Materials (Section B), in particular, we use the standard protocol defined in https://openaccess.thecvf.com/content_cvpr_2015/papers/Nie_Joint_Action_Recognition_2015_CVPR_paper.pdf to split the dataset.

**Comment by yuan0101213** [comment]
Thanks for your reply . I am trying to do the action recognition experiment in the paper!

**Comment by jenjsun** [comment]
No prob - also for bullet point 3 I updated it to say "the neg log of averaged matching probability". Hope everything goes well!

**Comment by yuan0101213** [comment]
Hi Team,

I am using the pr-vipe model to do action recognition experiment described by the paper(View-Invariant Probabilistic Embedding forHuman Pose) .

In the second step of your previous reply,we compute Pr-VIPE embeddings on the detected pose from each video frame in Penn Action. I get Penn Action and want to use infer.py(https://github.com/google-research/google-research/blob/master/poem/infer.py) to compute Pr-VIPE embeddings.

But how can I convert Penn Action to csv file described by the infer.py .

**Comment by jenjsun** [comment]
Hello, one important note is that you need to use the same keypoint detector on H3.6M to train Pr-VIPE as the keypoint detector on Penn Action to generate the 2D poses (which then can go into infer.py). If you followed the instructions on the Github, your Pr-VIPE model is trained using the Cascaded Pyramid Network (CPN) keypoints, so you would also need to run that detector on Penn Action. Alternatively, if you have your own 2D keypoint detector, you can run it on both H3.6M and Penn Action.

The other fields in the csv can be obtained by converting fields in the .mat file to .csv format, and this can be done using existing libraries in Python such as:
import scipy.io
mat = scipy.io.loadmat('file.mat')
Then you can write to csv once it is loaded. Hope your experiments go well!

**Comment by yuan0101213** [comment]
Hi,team,

Thanks for your reply. Now I get the result of infer.py with two video.
"embedding_stddevs.csv" , "unnormalized_embedding_samples.csv" , "unnormalized_embeddings.csv"

I read your paper in detail and know that you  get conclusion that the retrieved poses are close to their queries in the embedding space by computing the retrieval confidence of the model in section 4.2.3 . 

But I wonder how to compute the retrival confidence of  two frames by three csv files("embedding_stddevs.csv" , "unnormalized_embedding_samples.csv" , "unnormalized_embeddings.csv")

Sincerely.

**Comment by yuan0101213** [comment]
Thank you for your suggestion. I have finished the code work of the evaluation. 
However, every test sample needs to go through KNN and DTW algorithm with all train samples , which takes a long time. 
How long did it take you to complete action recognition. Is there any way to accelerate it?

**Comment by yuan0101213** [comment]
Thank you for your suggestion. I have finished the code work of the evaluation. 
However, every test sample needs to go through KNN and DTW algorithm with all train samples , which takes a long time. 
How long did it take you to complete action recognition. Is there any way to accelerate it?

**Comment by yuan0101213** [comment]
Thank you for your suggestion. I have finished the code work of the evaluation. 
However, every test sample needs to go through KNN and DTW algorithm with all train samples , which takes a long time. 
How long did it take you to complete action recognition. Is there any way to accelerate it?

**Comment by jenjsun** [comment]
Does your KNN implementation use parallel jobs? It seems that the sklearn implementation here (https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html) could be helpful. You could pre-compute the distance matrix and that would probably make it a lot faster. 
On our end it was ~25 minutes with parallelization.

**Comment by yuan0101213** [comment]
Thanks for your recommendation API.I use KNeighborsClassifier to train and predict.
knn = KNeighborsClassifier(n_neighbors=5,
                           algorithm='auto',
                           metric=cal_dtw_distance)
                          )
In the process of action recognition , each of my samples is a video. I reshape each sample (two dimension) into a one-dimensional matrix with the length of framenum * 16 (where 16 is embedding of frame). Because each video has different number of frames and the length of sample matrix is different, DTW algorithm is needed to align and calculate the distance of each sample.

But the train function(KNeighborsClassifier.fit(X_train, y_train))  needs samples of the same length. At present, I have two ideas to complement them:
1.Fill ∞ number, but the algorithm of kneighbors classifier （algorithm='auto'） will calculate the mean value of the ∞ number and the normal data of other samples when calculating the sample distance
2.Fill each sample into a one-dimensional index list, such as [videoID, frame_ start，frame_ end]。 When running the my distance function to calculate the distance, use the frame_ start，frame_ End find frames in embedding.csv File, get the matrix and calculate the distance. But it's going to be brute-force search again(algorithm='brute') ,which takes a long time..

How do you solve the problem that each training sample of kneighbors classifier model should be the same length (each sample is not the same length due to the number of frames)

**Comment by jenjsun** [comment]
Hello! For your first suggestion to fill in with inf, couldn't you handle by adding inside cal_dtw_distance to ignore inf frames? 
I discussed with co-authors and you could also break this into 2 steps:
1. compute the full dtw distance matrix. (If computation speed is still an issue here, you could try something like https://pypi.org/project/fastdtw/ (we didn't use this approximation, just a suggestion))
2. the pre-computed distance matrix should not too big, even running non-parallel knn search on this matrix should be relatively quick, and this step can also be parallelized by each query.

**Comment by farazBhatti** [comment]
@yuan0101213 , I am trying to do video alignment using DTW, it seems that you have implemented this already. Is it possible for you to share your implementation or any resources from where I can quickly figure out this algorithm? thanks...

**Comment by yuan0101213** [comment]
> @yuan0101213 , I am trying to do video alignment using DTW, it seems that you have implemented this already. Is it possible for you to share your implementation or any resources from where I can quickly figure out this algorithm? thanks...

You can go to Google to search the implementation of DTW algorithm, and then modify two places: 1. replace the data with the embedding of two videos; 2. replace the distance function with the distance function in the file(https://github.com/google-research/google-research/blob/master/poem/core/distance_utils.py).In this way, two video aligned frames and their similarity (distance) can be obtained.

Finally, the DTW blog I refer to is https://www.jianshu.com/p/05bee48cc6a2. Chinese VPN may be required.

**Comment by farazBhatti** [comment]
@yuan0101213 , thank you for quick response.
From blog this is the code for DTW:



```
from numpy import array, zeros, argmin, inf, equal, ndim
# from scipy.spatial.distance import cdist
from sklearn.metrics.pairwise import manhattan_distances
#在这里我用到的是曼哈顿距离(求绝对值距离)
#如果比较的是二维数组，则用欧几里得距离

s1 = [1, 2, 3, 4, 5, 5, 5, 4]
s2 = [3, 4, 5, 5, 5, 4]

r, c = len(s1), len(s2)
D0 = zeros((r+1,c+1))
D0[0,1:] = inf
D0[1:,0] = inf
D1 = D0[1:,1:]
#浅复制
# print D1

for i in range(r):
    for j in range(c):
        D1[i,j] = manhattan_distances(s1[i],s2[j])
#生成原始距离矩阵

M = D1.copy()
for i in range(r):
    for j in range(c):
        D1[i,j] += min(D0[i,j],D0[i,j+1],D0[i+1,j])
#代码核心，动态计算最短距离


i,j = array(D0.shape) - 2
#最短路径
# print i,j
p,q = [i],[j]
while(i>0 or j>0):
    tb = argmin((D0[i,j],D0[i,j+1],D0[i+1,j]))
    if tb==0 :
        i-=1
        j-=1
    elif tb==1 :
        i-=1
    else:
        j-=1
    p.insert(0,i)
    q.insert(0,j)

print M
#原始距离矩阵
print zip(p,q)
#匹配路径过程
print D1
#Cost Matrix或者叫累积距离矩阵
print D1[-1,-1]
#序列距离

```
What ive understood from your reply is that I should replace S1 and S2 with embeddings(320 dimension output by Pr-Vipe)
and replace distance function in the above code with distance function in distance_utils.py. But there is no distance function in the above code and there are multiple distance functions in distance_utlis.py? Can you please elaborate alittle bit more...
Thanks and much appreciated

**Comment by yuan0101213** [comment]
> @yuan0101213 , thank you for quick response. From blog this is the code for DTW:
> 
> ```
> from numpy import array, zeros, argmin, inf, equal, ndim
> # from scipy.spatial.distance import cdist
> from sklearn.metrics.pairwise import manhattan_distances
> #在这里我用到的是曼哈顿距离(求绝对值距离)
> #如果比较的是二维数组，则用欧几里得距离
> 
> s1 = [1, 2, 3, 4, 5, 5, 5, 4]
> s2 = [3, 4, 5, 5, 5, 4]
> 
> r, c = len(s1), len(s2)
> D0 = zeros((r+1,c+1))
> D0[0,1:] = inf
> D0[1:,0] = inf
> D1 = D0[1:,1:]
> #浅复制
> # print D1
> 
> for i in range(r):
>     for j in range(c):
>         D1[i,j] = manhattan_distances(s1[i],s2[j])
> #生成原始距离矩阵
> 
> M = D1.copy()
> for i in range(r):
>     for j in range(c):
>         D1[i,j] += min(D0[i,j],D0[i,j+1],D0[i+1,j])
> #代码核心，动态计算最短距离
> 
> 
> i,j = array(D0.shape) - 2
> #最短路径
> # print i,j
> p,q = [i],[j]
> while(i>0 or j>0):
>     tb = argmin((D0[i,j],D0[i,j+1],D0[i+1,j]))
>     if tb==0 :
>         i-=1
>         j-=1
>     elif tb==1 :
>         i-=1
>     else:
>         j-=1
>     p.insert(0,i)
>     q.insert(0,j)
> 
> print M
> #原始距离矩阵
> print zip(p,q)
> #匹配路径过程
> print D1
> #Cost Matrix或者叫累积距离矩阵
> print D1[-1,-1]
> #序列距离
> ```
> 
> What ive understood from your reply is that I should replace S1 and S2 with embeddings(320 dimension output by Pr-Vipe) and replace distance function in the above code with distance function in distance_utils.py. But there is no distance function in the above code and there are multiple distance functions in distance_utlis.py? Can you please elaborate alittle bit more... Thanks and much appreciated

According to the paper, I use compute_gaussian_likelihoods. The author also gave me some suggestions. Finally, I used acompute_gaussian_likelihoods.The following are the author's suggestions：
Hi Ducheng,

The `compute_gaussian_likelihoods` function is not what we used in the paper. I recommend referring to Eq 5 in our paper for how the matching probability p(m|z_i, z_j) between two embeddings z_i and z_j are computed. You can refer to this function for how the matching probability computed:
https://github.com/google-research/google-research/blob/678de1febc782cd76e75e9591fa3b5b9687b083c/poem/core/loss_utils.py#L70

Once you have the matching probability, the distance is simply -log p(m|z_i, z_j).

You can also directly use the L2 distance between the embedding means of the two predicted Gaussians ("unnormalized_embeddings.csv"), which should also work is much simpler.

Thanks,
Ting

**Comment by farazBhatti** [comment]
@yuan0101213 , thanks again for a thorough reply. I will apply your suggestions and update you here. Thanks

**Comment by B-rkh-verma** [comment]
Hi, I have one question how we can give our known 2D keypoints to the pr-vipe model.

**Comment by B-rkh-verma** [comment]
@yuan0101213 , Hi I want to ask how we can give our own keypoints to the pr-vipe model?
